src/main/java/fi/fmi/avi/archiver/config/factory/postaction/SwimRabbitMQPublisherFactory.java

```diff
@@ -43,19 +43,33 @@ public SwimRabbitMQPublisherFactory(
         this.clock = requireNonNull(clock, "clock");
     }
 
-    private static <T> T createLazyForwardingProxy(final Class<T> type, final Supplier<T> factory) {
-        final Supplier<T> delegate = Suppliers.memoize(() -> {
-            try {
-                return factory.get();
-            } catch (final RuntimeException exception) {
-                LOGGER.error("Could not create instance of {}: {}", type, exception.getMessage(), exception);
-                throw exception;
-            }
-        });
+    private static <T> T createLazyForwardingProxy(
+            final Class<T> type,
+            final Supplier<T> factory,
+            final AtomicReference<T> instanceRef) {
+        final Object lock = new Object();
         return type.cast(Proxy.newProxyInstance(
                 type.getClassLoader(),
-                new Class[]{type},
-                (proxy, method, args) -> method.invoke(delegate.get(), args)));
+                new Class<?>[]{type},
+                (proxy, method, args) -> {
+                    T instance = instanceRef.get();
+                    if (instance == null) {
+                        synchronized (lock) {
+                            instance = instanceRef.get();
+                            if (instance == null) {
+                                try {
+                                    instance = factory.get();
+                                    instanceRef.set(instance);
+                                } catch (final RuntimeException exception) {
+                                    LOGGER.error("Could not create instance of {}: {}", type, exception.getMessage(), exception);
+                                    throw exception;
+                                }
+                            }
+                        }
+                    }
```

Comment:
This is probably a theoretical problem, and it looks that in practice the code should work as expected. But I find using locks with `AtomicReference` a bit confusing, giving a false sense of security. If `instance == null`, I'd rather create a new instance and use `instanceRef.compareAndSet(null, instance)`. If operation returns `false`, then `close()` and abandon instance, logging and swallowing possible errors. In this case `T` needs to be changed to `T extends AutoCloseable`. This should probably be done in a `while` loop to ensure we finally have a proper `instance` to invoke.

I'm not sure if it makes the code any cleaner. But just something to consider and play with. I'm happy with the current solution too, since it is the only code setting a non-`null` value in the `instanceRef`.

Reply:
I made an attempt at the while loop solution. Quite hard to reason what is safe here.